AU2/plugins/sanity_checks/PlayerNotDead.py

```
@@ -0,0 +1,58 @@
+import re
+from typing import List, Dict, Container
+
+from AU2.database.model import Event
+from AU2.database.AssassinsDatabase import ASSASSINS_DATABASE
+from AU2.html_components import HTMLComponent
+from AU2.plugins.sanity_checks.model.SanityCheck import SanityCheck
+from AU2.plugins.sanity_checks.model.SanityCheck import Suggestion
+
+DX_NX_PATTERN = re.compile(r"(\[[DN](\d+)\])")
+
+class PlayerNotDead(SanityCheck):
+    """
+    Finds and detects cases where [DX] and [NX] appear,
+    despite the player being alive.
+    """
+
+    identifier = "Player_Not_Dead"
+
+    def _find_incorrect(self, string: str, dead_secret_ids: Container[str], fixes: Dict[str, str]):
+        # matches either [DX] or [NX]
+        for match in DX_NX_PATTERN.findall(string):
+            X = match[1]
+            if X not in dead_secret_ids:
+                fixes[match[0]] = X
+
+    def _gather_incorrect(self, e: Event):
+        dead_secret_ids = {ASSASSINS_DATABASE.get(victim_id)._secret_id for _, victim_id in e.kills}
+        output = {}
+        for (_, _, report) in e.reports:
+            self._find_incorrect(report, dead_secret_ids, output)
+        self._find_incorrect(e.headline, dead_secret_ids, output)
+        return output
+
+    def suggest_event_fixes(self, e: Event) -> List[Suggestion]:
+        suggestions = []
+        to_fix = self._gather_incorrect(e)
+        for original, secret_id in to_fix.items():
+            alist = ASSASSINS_DATABASE.get_filtered(include=lambda a: a._secret_id == secret_id,
+                                                    include_hidden=lambda a: a._secret_id == secret_id)
+            if len(alist) > 0:
+                assert len(alist) == 1

```

why not `if len(alist) == 1`? do we ever want to crash AU over a failed assert?

---

My thinking here was that with `len(alist) == 1` if we ever had `len(alist) > 1` then it would fail silently, and we would miss the fact that the database has somehow got screwed up.
*But* I suppose catching that here, in this sanity check, is not actually particularly useful seeing as it is unlikely to kick in before the duplication of ids becomes a problem.